File: typescript/leetcodehard/slidingWindowMaximum.ts
/*
https://leetcode.com/problems/sliding-window-maximum/

status - completed
*/

import { logp } from "../utils/logger";

function maxSlidingWindow(nums: number[], k: number): number[] {
  const deque: number[] = [];
  const result: number[] = [];

  for (let i = 0; i < nums.length; i++) {
    while (deque.length > 0 && deque[0] <= i - k) {
      deque.shift();
    }

    while (deque.length > 0 && nums[deque[deque.length - 1]] < nums[i]) {
      deque.pop();
    }

    deque.push(i);

    if (i >= k - 1) {
      result.push(nums[deque[0]]);
    }
  }

  return result;
}

logp(
  `input nums=[1,3,-1,-3,5,3,6,7] k=3 expected [3,3,5,5,6,7] actual ${JSON.stringify(
    maxSlidingWindow([1, 3, -1, -3, 5, 3, 6, 7], 3),
  )}`,
);
logp(`input nums=[1] k=1 expected [1] actual ${JSON.stringify(maxSlidingWindow([1], 1))}`);

File: python/leetcodehard/sliding_window_maximum.py
"""
https://leetcode.com/problems/sliding-window-maximum/

status - completed
"""

from collections import deque

from base_logger.logging_event import create_logger


class Solution:

    log = create_logger(__name__)

    def maxSlidingWindow(self, nums, k):
        window = deque()
        result = []

        for i, num in enumerate(nums):
            while window and nums[window[-1]] < num:
                window.pop()
            window.append(i)

            if window[0] <= i - k:
                window.popleft()

            if i >= k - 1:
                result.append(nums[window[0]])

        return result


if __name__ == "__main__":
    test_input_1 = ([1, 3, -1, -3, 5, 3, 6, 7], 3)
    output = Solution().maxSlidingWindow(*test_input_1)
    Solution().log.info("Output is %s", output)

    test_input_2 = ([1], 1)
    output = Solution().maxSlidingWindow(*test_input_2)
    Solution().log.info("Output is %s", output)
